contracts/staking/Staking.sol

```diff
@@ -0,0 +1,186 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import {ValueDistributor} from "./ValueDistributor.sol";
+
+contract Staking is ValueDistributor {
+    address private _sharesToken;
+    address private _rewardsToken;
+
+    /**
+     * @dev The rate of rewards distribution per second.
+     *
+     * It determines the rate at which rewards are earned and distributed
+     * to stakers based on their shares.
+     *
+     * Note: Ensure that the `_rate` value is set correctly to match
+     * the decimal precision of the `_rewardsToken` to ensure accurate rewards distribution.
+     */
+    uint256 private _rate;
+
+    uint256 private _stakingStartTime;
+
+    /**
+     * @dev Throws if the staking has not started yet.
+     */
+    modifier stakingStarted() {
+        _checkStakingStarted();
+
+        _;
+    }
+
+    /**
+     * @notice Initializes the contract setting the values provided by the deployer as shares token, rewards token, reward rate and staking start time.
+     * @param sharesToken_ The address of the shares token.
+     * @param rewardsToken_ The address of the rewards token.
+     * @param rate_ The reward rate.
+     * @param stakingStartTime_ The staking start time
+     */
+    constructor(
+        address sharesToken_,
+        address rewardsToken_,
+        uint256 rate_,
+        uint256 stakingStartTime_
+    ) {
+        require(sharesToken_ != address(0), "Staking: zero address cannot be the Shares Token");
+        require(rewardsToken_ != address(0), "Staking: zero address cannot be the Rewards Token");
+        require(rate_ > 0, "Rate has to be more than 0");
+
+        _sharesToken = sharesToken_;
+        _rewardsToken = rewardsToken_;
+        _setRate(rate_);
+        _setStakingStartTime(stakingStartTime_);
+    }
+
+    /**
+     * @notice Stakes the specified amount of tokens.
+     * @param amount_ The amount of tokens to stake.
+     */
+    function stake(uint256 amount_) public stakingStarted {
+        _addShares(msg.sender, amount_);
+    }
+
+    /**
+     * @notice Withdraws the specified amount of tokens.
+     * @param amount_ The amount of tokens to withdraw.
+     */
+    function withdraw(uint256 amount_) public stakingStarted {
```

Comment:
Let's rename this to `unstake`

Reply:
And add `withdraw` function that would call `unstake` and `claim` under the hood